OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,779 @@
+                           VkPipelineStageFlagBits pipeline_stage_flags) {
+    VkDevice device;
+    {
+      absl::ReaderMutexLock lock(&mutex_);
+      CHECK(command_buffer_to_device_.contains(command_buffer));
+      device = command_buffer_to_device_.at(command_buffer);
+    }
+
+    VkQueryPool query_pool = timer_query_pool_->GetQueryPool(device);
+
+    uint32_t slot_index;
+    bool found_slot = timer_query_pool_->NextReadyQuerySlot(device, &slot_index);
+    CHECK(found_slot);
+    dispatch_table_->CmdWriteTimestamp(command_buffer)(command_buffer, pipeline_stage_flags,
+                                                       query_pool, slot_index);
+
+    return slot_index;
+  }
+
+  std::vector<internal::QueueSubmission> PullCompletedSubmissions(VkDevice device,
+                                                                  VkQueryPool query_pool) {
+    std::vector<internal::QueueSubmission> completed_submissions = {};
+
+    absl::WriterMutexLock lock(&mutex_);
+    for (auto& [unused_queue, queue_submissions] : queue_to_submissions_) {
+      auto submission_it = queue_submissions.begin();
+      while (submission_it != queue_submissions.end()) {
+        const internal::QueueSubmission& submission = *submission_it;
+        if (submission.submit_infos.empty()) {
+          submission_it = queue_submissions.erase(submission_it);
+          continue;
+        }
+
+        bool erase_submission = true;
+        // Let's find the last command buffer in this submission, so first find the last
+        // submit info that has at least one command buffer.
+        // We test if for this command buffer, we already have a query result for its last slot
+        // and if so (or if the submission does not contain any command buffer) erase this
+        // submission.
+        auto submit_info_reverse_it = submission.submit_infos.rbegin();
+        while (submit_info_reverse_it != submission.submit_infos.rend()) {
+          const internal::SubmitInfo& submit_info = submission.submit_infos.back();
+          if (submit_info.command_buffers.empty()) {
+            ++submit_info_reverse_it;
+            continue;
+          }
+          // We found our last command buffer, so lets check if its result is there:
+          const internal::SubmittedCommandBuffer& last_command_buffer =
+              submit_info.command_buffers.back();
+          uint32_t check_slot_index_end = last_command_buffer.command_buffer_end_slot_index;
+
+          static constexpr VkDeviceSize kResultStride = sizeof(uint64_t);
+          uint64_t test_query_result = 0;
+          VkResult query_worked = dispatch_table_->GetQueryPoolResults(device)(
+              device, query_pool, check_slot_index_end, 1, sizeof(test_query_result),
+              &test_query_result, kResultStride, VK_QUERY_RESULT_64_BIT);
+
+          // Only erase the submission if we query its timers now.
+          if (query_worked == VK_SUCCESS) {
+            erase_submission = true;
```

Comment:
This reassignment here confuses me a bit as `erase_submission` was already assigned true outside of the loop. But maybe it would be less clear without the reassignment?

Reply:
I am not sure, I had the fear, that at that point, the reader don't has this in mind. I'd keep it, if you are not strongly against it.